docs: concepts, getting started and programming reference

eeberhard opened the pull request:
<!-- Pull Request guidelines:
1. Give the PR a relevant and descriptive title, using one of the supported commit types:
   [ build, ci, chore, docs, feat, fix, perf, refactor, release, revert, style, test ]
2. Link the PR to the parent issue, which should already describe and motivate the PR
3. Explain how the PR addresses the parent issue
4. Add any supporting resources (screenshots, test results, etc)
5. Help the reviewer by suggesting the method and estimated time to review
6. If applicable, make a checklist of tasks that should be completed before merging
7. If applicable, mention related issues (blocked by / blocks)
8. Post creation actions: tag reviewers and link the PR to its parent issue under the Development section
-->

## Description

<!-- Required: link the parent issue -->

I have been busy over the last weeks chipping away at our documentation. It took a lot of time and it's certainly not complete, but I think it's worth it to have the core concepts and references in one place.

I'm bringing the whole gang together to review this one because I will be happy to have as many eyes on it as possible! 

## Viewing the documentation

If you have `npm` installed, you can check out this repository and go to the `docs` directory in a terminal. Then run:

```bash
npm install
npm start
```
The docs will be viewable on `localhost:3000`.

Otherwise, I have attached [the generated documentation build output as a zip file](https://github.com/aica-technology/api/files/12504841/build.zip). You can unzip the the build folder and just serve it with any HTTP server. For example: 



```bash
cd build
python3 -m http.server
```
In this case, the docs will be viewable on `localhost:8000`.

## Review guidelines

<!-- Required: estimate how long a review should take -->
Estimated Time of Review: 180 minutes

I would be grateful if you could review:

- High-level structure of guides, pages and sections
- Clarity and impact of content
- General language and writing style
- Nitpicks: typos, missing links, etc

Most importantly, please keep an open mind about what is missing, what is not clear or anything I am wrong about!

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

<!-- Optional: define further tasks that should be completed before merging
- [x] Task 1
- [ ] Task 2
-->

<!-- Optional: link related issues
## Related issues
### Blocked by:
- #0

### Blocks:
- #0
-->

LouisBrunner commented on docs/docusaurus.config.js:
```diff
@@ -97,13 +113,12 @@ const config = {
         style: "dark",
         links: [
-          // TODO: Example of how to add an entry to the docs
           // {
```
If it's unused, we can just delete it?

LouisBrunner commented on docs/docs/reference/02-yaml-syntax.md:
```diff
@@ -0,0 +1,638 @@
+## Validating a YAML Application
+
+<!-- FIXME: link to the schema on GitHub once it is on main; relative paths will break if the doc is versioned -->
+The [YAML application schema](../../../schemas/applications/schema/application.schema.json) defines the structural rules
```
Does it work at all when deployed on GitHub?

eeberhard replied:
Probably not. I want to release to main after this anyway, which will involve updating and tagging the schema version at which point it can be permalinked properly here. So it's a TODO but will be resolved soon.

LouisBrunner commented on docs/docs/getting-started/03-run.md:
```diff
@@ -0,0 +1,77 @@
+
+# Running the image
+
+This guide assumes that the AICA software image was distributed as `aica-technology/base`. If the base image has a
```
Should we change the default name to `registry.licensing.aica.tech/base`?

eeberhard replied:
Since that is our registry URL and we will allow user login with `docker pull`, it would make sense.

But, the `base` image pulled from ECR is not runnable directly because it needs still to be packaged, right? (creation of manifest happens in the packing script). We package and tag it manually for now, so I would prefer to keep a "simple" readable name. Then we can update this once we have completed registry access and packaging on the user-side.

LouisBrunner commented on docs/docs/getting-started/01-intro.md:
```diff
@@ -0,0 +1,20 @@
+
+The AICA framework is being distributed in a limited beta program. Installation, configuration and run-time instructions
+are subject to breaking changes in the near future. We will endeavour to keep this documentation up-to-date with the
+latest versions. In case of doubt, contact support@aica.tech for help.
```
Did this email get setup already?

eeberhard replied:
Yes, we have `support@aica.tech` intended for existing users / support questions, and `contact@aica.tech` for all general and sales enquiries. I have tried to use the most appropriate address whenever it comes up

LouisBrunner commented on docs/docs/concepts/06-going-further.md:
```diff
@@ -0,0 +1,12 @@
+
+# Going further with AICA
+
+[Contact us](mailto:contact@aica.tech) to learn more about AICA software or and to request software license.
```
Different emails than before FYI

domire8 commented on docs/docs/concepts/04-aica-applications.md:
```diff
@@ -0,0 +1,68 @@
+To address this challenge, the AICA framework includes at its core a generalized execution process known as the dynamic
+state engine.
+
+## The dynamic state engine
```
IMO the state engine deserves to be capitalized, just like the controller manager: The Dynamic State Engine

domire8 commented on docs/docs/concepts/04-aica-applications.md:
```diff
@@ -0,0 +1,68 @@
+The dynamic state engine manages the application state using **events**, which are used to reconfigure parameters, load
+or unload components, manage component lifecycle states, trigger specific service actions and more.
+
+## Declarative application descriptions
```
nitpick: Even though it's the correct word, I find 'declarative' a bit heavy in this context. Is an adjective even needed?

eeberhard replied:
To me, declarative is very important in this context. It refers to the programming paradigm, and the implementation / program flow is abstract.

I also wouldn't want to have just `## Application descriptions` or `## Application syntax` on its own as a section header.

I prefer the term "application descriptions" which can include both YAML and graph forms. Application syntax refers more specifically to the YAML. But, both the visual and scripted descriptions are forms of declarative programming.

domire8 commented on docs/docs/concepts/05-building-blocks/01-signals.md:
```diff
@@ -0,0 +1,88 @@
+graph without modifying or recompiling any source code. By using standard message types, the signal compatibility
+between components and controllers is also simplified.
+
+<!-- TODO: not sure if this is too much to put here, since it's more relevant to the programming reference
```
Hmm I do think that this could be relevant, maybe in combination with a rewrite of the State messages section below (see comment).

domire8 commented on docs/docs/concepts/05-building-blocks/01-signals.md:
```diff
@@ -0,0 +1,88 @@
+
+### State messages
+
+In robot control applications, the _state_ of a robot or other objects is highly important.
```
nitpick: Again, there is nothing incorrect with the way you wrote it but here, I got the feeling that the state message type was not highlighted enough. When I read it, I saw the list of standard messages, then you see the state types from state representation but the `make it easy ... state message types` is in my opinion not prominent enough and doesn't highlight the fact that this is (probably) the main topic type and a big simplification.

domire8 commented on docs/docs/concepts/05-building-blocks/02-events.md:
```diff
@@ -0,0 +1,69 @@
+# Events
+
+Events are discrete actions that change the dynamic state of an AICA application. They are handled and executed by the
+dynamic state engine For this reason, they are sometimes referred to as "state events".
```
Missing full stop after state engine

domire8 commented on docs/docs/concepts/05-building-blocks/02-events.md:
```diff
@@ -0,0 +1,69 @@
+## Triggering events
+
+Events can be triggered externally through interactions with the Developer Interface UI or via the REST API, or
+internally by component predicates and conditions as defined in the application description.
```
Here, you refer to predicates just before you explain what they are. Some people might already get worried that they don't know what predicates are. Should me maybe move the following paragraph to the top of the page? Predicates are quite core to the events so having that section up top might not hurt

eeberhard replied:
I understand your point but it is really an introduction / transition sentence to the bold section header below. I'll try highlighting / rephrasing it to make it clearer.

domire8 commented on docs/docs/concepts/05-building-blocks/05-hardware-interfaces.md:
```diff
@@ -0,0 +1,17 @@
+sidebar_position: 5
+---
+
+# Hardware Interfaces
```
I believe 'Hardware interfaces' was not capitalized earlier. Shouldn't it be 'Hardware interfaces' then?

LouisBrunner reviewed (approved):
Bunch of various questions, none of them blocking

domire8 reviewed (commented):
I reviewed the first section and will continue later with a fresh mind.

This looks already super good, thanks for putting this together! I have a few comments but they are not blocking, just some impressions I had.

One thing I would like to add is that the controllers and hardware interfaces sections at the end of concepts seem a bit short compared to the rest which makes you think you ran out of energy. I guess it's just not easy to strike a balance of being high level and going into detail for these concepts.

jjenscodee commented on docs/docs/reference/02-yaml-syntax.md:
````diff
@@ -0,0 +1,638 @@
+  ...
+```
+
+## On Start
````
Do we need to update the docs for yaml syntax? for example, we have buttons now in the application.yml

eeberhard replied:
Good point, buttons are not documented. Because buttons are a graph / UI feature, and this is a programming reference for developers who might just want to use the YAML syntax on its own, I will leave that as a TODO.

jjenscodee commented on docs/docs/getting-started/04-timer-example.md:
````diff
@@ -0,0 +1,143 @@
+  load: timer
+```
+
+The graph should show two components connected with event edges. Press the Play button to start the application.
````
Do you think if it looks better if we also include a screenshot of the graph as an example?

eeberhard replied:
Screenshots / videos (or gifs) would certainly be nice. I was reluctant to include those at first because they may be outdated sooner than a text description would be, but I guess better to have something than nothing.

jjenscodee commented on docs/docs/getting-started/04-timer-example.md:
```diff
@@ -0,0 +1,143 @@
+
+## Putting it all together
+
+When the application is started, the `timer` component is loaded. It is initially unconfigured, which triggers it
```
Maybe we can put a video here? A video that runs this example with the state turns red/green

eeberhard replied:
I think an animated gif of that would be nice. I'll try and get that done.

jjenscodee reviewed (approved):
Looks great!

Just a small question: Why having this "Edit this page" on every page? What is this for?
<img width="168" alt="Screenshot 2023-09-04 at 18 24 11" src="https://github.com/aica-technology/api/assets/71389754/434b6843-eca5-4783-97ec-fe508271f13c">

JasperTan97 commented on docs/docs/reference/02-yaml-syntax.md:
````diff
@@ -0,0 +1,638 @@
+    force_torque_sensor: "/force"
+```
+
+#### Component period
````
This is my first comment, so let me know if I am doing it wrong. 
Suggestion: component period would do better directly below Parameters as opposed to Inputs and Outputs.

domire8 commented on docs/docs/reference/02-yaml-syntax.md:
````diff
@@ -0,0 +1,638 @@
+    y: 200
+```
+
+### Log Level
````
Here, a bit of inconsistency with capitalizing (Log Level vs Display name)

domire8 commented on docs/docs/reference/02-yaml-syntax.md:
```diff
@@ -0,0 +1,638 @@
+
+:::caution
+
+All hardware interfaces in the application are automatically loaded and initialized when the application starts.
```
Wait is that true or do you want it to be like that?

eeberhard replied:
Well yes that is essentially true at the moment (auto-load hardware has always been default true in all of our use-cases). The service and hardware refactor will give us a bit more fine-grained control over initial and subsequent hardware states in future, but for now this is a fair summary.

domire8 commented on docs/docs/reference/custom-components/01-component-package.md:
```diff
@@ -0,0 +1,135 @@
+include_directories(include)
+
+# highlight-start
+### Register and install C++ Components here ###
```
Can you control the color of the highlight? Black font on grey background is a bit hard to read

eeberhard replied:
Yeah it looks good in dark mode but quite bad in light mode. I'll see if I can adjust it, since the ability to highlight a line is still quite nice.

domire8 commented on docs/docs/reference/custom-components/02-custom-component.md:
```diff
@@ -0,0 +1,647 @@
+                return False
+        elif parameter.get_name() == "B":
+            if parameter and parameter.get_value() < 0.0:
+                self.get_logger().warn(f"Parameter B cannot be negative ({parmaeter.get_value()}). "
```
```suggestion
                self.get_logger().warn(f"Parameter B cannot be negative ({parameter.get_value()}). "
```

domire8 commented on docs/docs/reference/custom-components/02-custom-component.md:
```diff
@@ -0,0 +1,647 @@
+
+The supported message types are defined in the `std_msgs` library and include:
+
+- Boolean
```
Is this capital B intended?

eeberhard replied:
Yes, it's like Cartesian, named after some old dude.

domire8 commented on docs/docs/reference/custom-components/02-custom-component.md:
````diff
@@ -0,0 +1,647 @@
+<TabItem value="c++" label="C++">
+
+```cpp title="include/custom_component_package/CppComponent.hpp"
+class CppComponent : public modulo_components::Component {
````
This is not correct, is it? Components don't have an on_step_callback. https://github.com/aica-technology/modulo/blob/main/source/modulo_components/src/Component.cpp

eeberhard replied:
I wrote this preemptively because I think they should, I was planning to tag you in this when you got back.
https://github.com/aica-technology/modulo/issues/49

domire8 commented on docs/docs/reference/custom-components/03-component-descriptions.md:
```diff
@@ -0,0 +1,227 @@
+<!-- FIXME: link to the schema on GitHub once it is on main; relative paths will break if the doc is versioned -->
+The elements of a component description are written as a standardized JSON file for each component. The expected
+structure of the description is defined by the JSON schema
+in [component.schema.json](../../../../schemas/component-descriptions/schema/component.schema.json)
```
Shouldn't that be something like 'Component Schema' as link name?

domire8 reviewed (commented):
Finished my review now, a bunch of small comments. Really great docs overall, thanks for that!

eeberhard commented:
> Why having this "Edit this page" on every page?

@jjenscodee this is a convenience for "live" documentation. We have markdown files in GitHub that are built and deployed through docusaurus as documentation pages. This edit link is basically a shortcut to get to the original markdown file on GitHub from the deployed documentation page. That way, if you see a typo or something else that needs changing, you can easily find the source file. It also invites open-source documentation contributions since you can make a fork and open a PR if you want to edit the page.

eeberhard commented on docs/docs/getting-started/04-timer-example.md:
```diff
@@ -0,0 +1,143 @@
+
+# A basic application example
+
+Open the Developer Interface and copy the following application code into the Editor.
```
Yes because it is the title of the tab (capitalized in the UI). In general otherwise it is lower-case. I will rewrite to clarify.

eeberhard commented on docs/docs/reference/02-yaml-syntax.md:
```diff
@@ -0,0 +1,638 @@
+hardware_interfaces:
+  robot_a:
+    urdf: ...
+    rate: ...
```
Yes, it's a fair point. The name `period` was held over from the modulo `Cell` days because it was the input to the wall timer function (a period for the timer to wait). Conversely the `rate` term was adopted from the ros2_control stack, where control engineers like to think about frequencies. 

I don't know which one of period or frequency/rate is more intuitive, but mixing the two in one syntax is not ideal. A `period` for periodic behavior is quite nice conceptually, but I would also not be against components having a rate instead. Only problem is that it would be a breaking change unless we support both parameters, and in that case we would need to decide which one to prioritize if both are set.

You could open a corresponding issue in modulo if you want.

domire8 replied:
For me, a frequency is more intuitive because you can ask yourself how many times a second should it execute which in my opinion is easier for the brain to imagine that saying how long on period should be

domire8 replied:
Opened it :+1:

eeberhard commented:
Thanks for your thorough reviews!

I think I addressed all the comments, except for Dominic's first point (_"the controllers and hardware interfaces sections at the end of concepts seem a bit short"_). It's true I ran out of energy a bit while writing that, but I'm not sure what to add there and I don't want to bloat it unnecessarily.

domire8 reviewed (approved):
Great great work! thanks for that.

(and sorry about the mess with the wrong issue :smile:)
